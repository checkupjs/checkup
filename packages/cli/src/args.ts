export const pathArg = {
  name: 'paths',
  description:
    'The paths or globs that checkup will operate on. If no paths or globs are provided, checkup will run on the entire directory beginning at --cwd.',
};
