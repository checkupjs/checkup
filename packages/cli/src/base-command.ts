import { CheckupError } from '@checkup/core';
import { Command } from '@oclif/command';
import { red } from 'chalk';

export abstract class BaseCommand extends Command {
  extendedError(
    error: string | Error | CheckupError,
    options: { code?: string; exit?: number | false } = {}
  ) {
    if (typeof error === 'string') {
      error = new Error(error);
    }

    if (options.exit === false) {
      console.error(
        error instanceof CheckupError ? error.render() : `\n${red('Error')} ${error.message}`
      );
    } else {
      throw error;
    }
  }
}
