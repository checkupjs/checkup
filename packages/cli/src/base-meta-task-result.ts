import { TaskIdentifier } from '@checkup/core';

export default abstract class BaseMetaTaskResult {
  meta: TaskIdentifier;

  constructor(meta: TaskIdentifier) {
    this.meta = meta;
  }
}
