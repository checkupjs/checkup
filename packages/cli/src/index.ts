import * as path from 'path';

import { Command, flags } from '@oclif/command';
import {
  TaskConstructor,
  TaskList,
  TaskResult,
  getPackageJson,
  getRegisteredTasks,
  ui,
} from '@checkup/core';

function mergeTaskResults(taskResults: TaskResult[]) {
  let mergedResults = {};

  taskResults.forEach(taskResult => {
    mergedResults = Object.assign(mergedResults, taskResult.toJson());
  });

  return mergedResults;
}

class Checkup extends Command {
  static description = 'A CLI that provides health check information about your project';

  static args = [
    {
      name: 'path',
      required: true,
      description: 'The path referring to the root directory that Checkup will run in',
      default: '.',
    },
  ];

  static flags = {
    version: flags.version({ char: 'v' }),
    help: flags.help({ char: 'h' }),
    force: flags.boolean({ char: 'f' }),
    silent: flags.boolean({ char: 's' }),
    json: flags.boolean(),
    task: flags.string({ char: 't' }),
  };

  async run() {
    let { args, flags } = this.parse(Checkup);
    let registeredTasks: Map<string, TaskConstructor>;

    try {
      getPackageJson(args.path);
    } catch (e) {
      this.error(
        `The ${path.resolve(
          args.path
        )} directory found through the 'path' option does not contain a package.json file. You must run checkup in a directory with a package.json file.`,
        e
      );
    }

    await this.config.runHook('register-tasks', {});

    registeredTasks = getRegisteredTasks();

    let taskResults;
    let tasksToBeRun = new TaskList();

    if (flags.task !== undefined) {
      let task = registeredTasks.get(flags.task);

      if (task !== undefined) {
        tasksToBeRun.addTask(task, args);
      }
    } else {
      tasksToBeRun.addTasks(Array.from(registeredTasks.values()), args);
    }

    ui.action.start('Checking up on your project');

    taskResults = await tasksToBeRun.runTasks();

    if (!flags.silent) {
      if (flags.json) {
        ui.styledJSON(mergeTaskResults(taskResults));
      } else {
        taskResults.forEach(taskResult => taskResult.toConsole());
      }
    }

    ui.action.stop();
  }
}

export = Checkup;
