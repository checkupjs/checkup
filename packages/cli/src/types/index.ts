export default {};

export const enum ProjectType {
  App = 'application',
  Addon = 'addon',
  Engine = 'engine',
  Unknown = 'unknown',
}

export const enum TestType {
  Application = 'application',
  Container = 'container',
  Rendering = 'rendering',
  Unit = 'unit',
}

export interface Task {
  run: () => Promise<TaskResult>;
}

export interface TaskConstructor {
  new (): Task;
}

export interface TaskItemData {
  type: string;
  data: string[];
  total: number;
}

export interface TaskResult {
  toConsole: () => void;
  toJson: () => {};
}

export type SearchPatterns = Record<string, string[]>;
