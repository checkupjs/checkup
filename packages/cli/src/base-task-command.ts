import { Result, Log, Invocation } from 'sarif';
import {
  RunFlags,
  TaskError,
  Action,
  CheckupConfig,
  Task,
  TaskContext,
  ui,
  CheckupError,
  getRegisteredActions,
  getConfigPathFromOptions,
  getConfigPath,
  readConfig,
  registerParser,
  registerActions,
  registerTaskReporter,
  loadPlugins,
  getRegisteredParsers,
  FilePathArray,
  getFilePathsAsync,
  buildNotificationsFromTaskErrors,
} from '@checkup/core';
import { BaseCommand } from './base-command';
import TaskList from './task-list';
import { getLog } from './get-log';
import { TASK_ERRORS } from './task-errors';
import { getPackageJson } from './utils/get-package-json';
import { getReporter } from './reporters/get-reporter';
import { reportAvailableTasks } from './reporters/available-task-reporter';

let __tasksForTesting: Set<Task> = new Set<Task>();

export function _registerTaskForTesting(task: Task) {
  __tasksForTesting.add(task);
}
export function _resetTasksForTesting() {
  __tasksForTesting = new Set<Task>();
}

export abstract class BaseTaskCommand extends BaseCommand {
  runArgs!: string[];
  runFlags!: RunFlags;
  pluginTasks: TaskList = new TaskList();
  pluginTaskResults: Result[] = [];
  pluginTaskErrors: TaskError[] = [];
  startTime: string = '';
  actions: Action[] = [];
  checkupConfig!: CheckupConfig;
  executedTasks!: Task[];
  cliModeEnabled: boolean = true;
  taskContext!: TaskContext;

  get taskFilterType() {
    if (this.runFlags.task !== undefined) {
      return 'task';
    } else if (this.runFlags.category !== undefined) {
      return 'category';
    } else if (this.runFlags.group !== undefined) {
      return 'group';
    }

    return '';
  }

  public async run() {
    await this.loadConfig();

    await this.register();

    if (this.runFlags['list-tasks']) {
      this.printAvailableTasks();
    } else {
      if (this.cliModeEnabled) {
        ui.action.start('Checking up on your project');
      }

      await this.runTasks();
      await this.runActions();

      let log: Log = await getLog(
        this.taskContext,
        this.pluginTaskResults,
        this.actions,
        this.getInvocation(this.pluginTaskErrors),
        this.pluginTasks,
        this.executedTasks
      );

      if (this.cliModeEnabled) {
        this.report(log);
        ui.action.stop();
      }

      return log;
    }
  }

  private async runTasks() {
    if (this.taskFilterType) {
      let { tasksFound, tasksNotFound } = this.findTasks();

      if (tasksFound.length > 0) {
        [this.pluginTaskResults, this.pluginTaskErrors] = await this.pluginTasks.runTasks(
          tasksFound
        );
      }

      if (tasksNotFound.length > 0) {
        let error = TASK_ERRORS.get(this.taskFilterType)!;

        this.extendedError(new CheckupError(error.message(tasksNotFound), error.callToAction));
        ui.action.stop();
      }

      this.executedTasks = tasksFound;
    } else {
      [this.pluginTaskResults, this.pluginTaskErrors] = await this.pluginTasks.runTasks();
      this.executedTasks = this.pluginTasks.getTasks();
    }
  }

  private findTasks() {
    if (this.runFlags.task !== undefined) {
      return this.pluginTasks.findAllByTaskName(...this.runFlags.task);
    } else if (this.runFlags.category !== undefined) {
      return this.pluginTasks.findAllByCategory(...this.runFlags.category);
    } else if (this.runFlags.group !== undefined) {
      return this.pluginTasks.findAllByGroup(...this.runFlags.group);
    }

    return { tasksFound: [], tasksNotFound: [] };
  }

  private runActions() {
    return new Promise<void>((resolve, reject) => {
      let evaluators = getRegisteredActions();

      for (let [taskName, evaluator] of evaluators) {
        let task = this.pluginTasks.find(taskName);

        let taskResult = this.pluginTaskResults.filter((result: Result) => {
          return result.properties?.taskName === taskName;
        });

        if (task && taskResult.length > 0) {
          try {
            this.actions.push(...evaluator(taskResult, task.config));
          } catch (error) {
            reject(error);
          }
        }
      }

      resolve();
    });
  }

  private async loadConfig() {
    let configPath;

    try {
      configPath =
        (await getConfigPathFromOptions(this.runFlags.config)) || getConfigPath(this.runFlags.cwd);
      this.checkupConfig = readConfig(configPath);

      let plugins = await loadPlugins(this.checkupConfig.plugins, this.runFlags.cwd);

      this.config.plugins.push(...plugins);
    } catch (error) {
      this.extendedError(error);
    }
  }

  protected async register() {
    await this.config.runHook('register-parsers', {
      registerParser,
    });

    await this.config.runHook('register-actions', {
      registerActions,
    });

    await this.config.runHook('register-task-reporter', {
      registerTaskReporter,
    });

    // if excludePaths are provided both via the command line and config, the command line is prioritized
    let excludePaths = this.runFlags['exclude-paths'] || this.checkupConfig.excludePaths;

    this.taskContext = Object.freeze({
      cliArguments: this.runArgs,
      cliFlags: this.runFlags,
      parsers: getRegisteredParsers(),
      config: this.checkupConfig,
      pkg: getPackageJson(this.runFlags.cwd),
      paths: FilePathArray.from(
        await getFilePathsAsync(this.runFlags.cwd, this.runArgs, excludePaths)
      ) as FilePathArray,
    });

    this.registerTasks();

    __tasksForTesting.forEach((task: Task) => {
      this.pluginTasks.registerTask(task);
    });
  }

  protected abstract registerTasks(): void;

  private report(log: Log) {
    let generateReport = getReporter(this.runFlags);

    generateReport(log, this.runFlags);
  }

  private printAvailableTasks() {
    reportAvailableTasks(this.pluginTasks);
  }

  private getInvocation(errors: TaskError[]): Invocation {
    return {
      arguments: this.runArgs,
      executionSuccessful: true,
      endTimeUtc: new Date().toJSON(),
      environmentVariables: {
        cwd: this.runFlags.cwd,
        outputFile: this.runFlags['output-file'],
        format: this.runFlags.format,
      },
      toolExecutionNotifications: buildNotificationsFromTaskErrors(errors),
      startTimeUtc: this.startTime,
    };
  }
}
