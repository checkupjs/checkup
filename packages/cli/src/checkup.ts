import cli from 'cli-ux';
import { ITaskConstructor, ITaskResult } from './types';
import { getTaskByName } from './utils/default-tasks';
import TaskList from './task-list';
import * as DefaultTasks from './tasks';
import ResultWriter from './utils/result-writer';
import Clock from './utils/clock';

const DEFAULT_TASKS = <ITaskConstructor[]>(
  Object.values(DefaultTasks).filter(x => typeof x == 'function')
);

/**
 * @class Checkup
 *
 * The entry point for invoking all checkup tasks.
 */
export default class Checkup {
  args: any;
  flags: any;
  defaultTasks: ITaskConstructor[];

  /**
   *
   * @param project {IProject} the project model that is instantiated as part of ember-cli.
   * @param ui {IUserInterface} the UI model that is instantiated as part of ember-cli.
   */
  constructor(args: any, flags: any, tasks: ITaskConstructor[] = DEFAULT_TASKS) {
    this.args = args;
    this.flags = flags;
    this.defaultTasks = tasks;
  }

  /**
   * @method run
   *
   * Gathers and runs all tasks associated with checking up on an Ember repo.
   */
  async run(): Promise<ITaskResult[]> {
    let clock = new Clock();
    let tasks = new TaskList();

    if (this.flags.task !== undefined) {
      let task = getTaskByName(this.flags.task);

      tasks.addTask(task);
    } else {
      tasks.addTasks(this.defaultTasks);
    }

    clock.start();

    cli.action.start('Hang tight while we check up on your Ember project');

    let taskResults = await tasks.runTasks();

    clock.stop();

    cli.action.stop();

    if (!this.flags.silent) {
      let writer = new ResultWriter(taskResults);

      if (this.flags.json) {
        console.log(JSON.stringify(writer.toJson(), null, 2));
      } else {
        writer.toConsole();
      }
      writer.writeDuration(clock.duration);
    }

    return taskResults;
  }
}
