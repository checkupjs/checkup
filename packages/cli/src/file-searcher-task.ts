import { SearchPatterns, Task, TaskResult } from './types';

import { BASE_DIR } from './utils/project';
import FileSearcher from './searchers/file-searcher';

/**
 * @class FileSearcherTask
 * @extends Task
 * @implements ITask
 *
 * A checkup task specific to file searcher used to encapsulate an operation that
 * checks certain characteristics of your Ember project.
 */
export default abstract class FileSearcherTask implements Task {
  searcher: FileSearcher;

  /**
   *
   * @param result {TaskResult[]} the result object that aggregates data together for output.
   * @param searchPatterns {SearchPatterns} the search pattern that your filesearcher uses to return the results.
   */
  constructor(searchPatterns: SearchPatterns) {
    this.searcher = new FileSearcher(BASE_DIR, searchPatterns);
  }

  abstract run(): Promise<TaskResult>;
}
