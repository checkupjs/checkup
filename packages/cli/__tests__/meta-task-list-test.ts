import { MetaTask, MetaTaskResult } from '../src/types';

import MetaTaskList from '../src/meta-task-list';
import MockMetaTaskResult from './__utils__/mock-meta-task-result';

class FakeMetaTask implements MetaTask {
  meta = {
    taskName: 'fake-meta-task',
    friendlyTaskName: 'Fake Meta Task',
  };

  async run(): Promise<MetaTaskResult> {
    return new MockMetaTaskResult(this.meta, 'fake meta task is being run');
  }
}

class OtherFakeMetaTask implements MetaTask {
  meta = {
    taskName: 'other-fake-meta-task',
    friendlyTaskName: 'Other Fake Meta Task',
  };

  async run(): Promise<MetaTaskResult> {
    return new MockMetaTaskResult(this.meta, 'other fake meta task is being run');
  }
}

describe('MetaTaskList', () => {
  it('can create an instance of a MetaTaskList', () => {
    let taskList = new MetaTaskList();

    expect(taskList).toBeInstanceOf(MetaTaskList);
  });

  it('hasTask returns false if no task exists with that name', () => {
    let taskList = new MetaTaskList();

    taskList.registerTask(new FakeMetaTask());

    expect(taskList.hasTask('foo')).toEqual(false);
  });

  it('hasTask returns true if task exists with that name', () => {
    let taskList = new MetaTaskList();

    taskList.registerTask(new FakeMetaTask());

    expect(taskList.hasTask('fake-meta-task')).toEqual(true);
  });

  it('findTask returns undefined if no task exists with that name', () => {
    let taskList = new MetaTaskList();

    taskList.registerTask(new FakeMetaTask());

    expect(taskList.findTask('foo')).toBeUndefined();
  });

  it('findTask returns task instance if task exists with that name', () => {
    let taskList = new MetaTaskList();

    taskList.registerTask(new FakeMetaTask());

    expect(taskList.findTask('fake-meta-task')).toBeDefined();
  });

  it('runTask will run a task by taskName', async () => {
    let taskList = new MetaTaskList();

    taskList.registerTask(new FakeMetaTask());

    let result = await taskList.runTask('fake-meta-task');

    expect(result.json()).toMatchInlineSnapshot(`
      Object {
        "fake-meta-task": "fake meta task is being run",
      }
    `);
  });

  it('runTasks will run all registered tasks', async () => {
    let taskList = new MetaTaskList();

    taskList.registerTask(new FakeMetaTask());
    taskList.registerTask(new OtherFakeMetaTask());

    let result = await taskList.runTasks();

    expect(result[0].json()).toMatchSnapshot();
    expect(result[1].json()).toMatchSnapshot();
  });
});
