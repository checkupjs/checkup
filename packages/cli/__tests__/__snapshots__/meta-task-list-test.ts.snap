// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MetaTaskList runTasks will run all registered tasks 1`] = `
Object {
  "fake-meta-task": "fake meta task is being run",
}
`;

exports[`MetaTaskList runTasks will run all registered tasks 2`] = `
Object {
  "other-fake-meta-task": "other fake meta task is being run",
}
`;
